### PR TITLE
Time-to-close and further memory improvements

### PR DESCRIPTION
**Pull request by dkerr64**
fixes for issues #60 (time-to-close delay) and #143 (only reboot for settings change when necessary).  Also includes cleanup and removal of unnecessary code / html pages.

Added implementation for #150

### CODE REVIEW

**Comment by jgstroud** [comment]
at @dkerr64 can we move that last commit for the OTA to a separate PR?

**Comment by dkerr64** [comment]
> at @dkerr64 can we move that last commit for the OTA to a separate PR?

Yes, done. PR #153

File: src/web.cpp [hunk 1]
```diff
@@ -436,12 +443,11 @@ void handle_status()
     ADD_INT(json, "TTCseconds", TTCdelay);
     END_JSON(json);
 
-    RINFO("Status requested:\n%s", json);
+    RINFO("Status requested:\n%s\nlength %d\n", json, strlen(json));
```
Comment (jgstroud) [inline src/web.cpp]: You don't need to keep this change, I was just curious how close we were to filling up that string

File: src/web.cpp [hunk 2]
```diff
@@ -94,7 +96,7 @@ int crashCount = 0;
 // For Server Sent Events (SSE) support
 // Just reloading page causes register on new channel.  So we need a reasonable number
 // to accommodate "extra" until old one is detected as disconnected.
-#define SSE_MAX_CHANNELS 8
+#define SSE_MAX_CHANNELS 4
```
Comment (jgstroud) [inline src/web.cpp:106]: I tried moving this to the iram heap as well, but it cratered.  not sure why, so I just reduced to 4.

File: src/comms.cpp [hunk 3]
```diff
@@ -859,6 +892,19 @@ void open_door() {
     door_command(DoorAction::Open);
 }
 
+void TTCdelayLoop() {
+    if (--TTCcountdown > 0) {
```
[the remark this answers is not in the document]
Comment (dkerr64) [reply src/comms.cpp:896]: > There is a "service_timers_loop" in the ratgdo.cpp file. Should this go there instead?

I probably erred in naming this xxxLoop as this function is not part of the main loop.  It executes once every 500ms using Ticker().  And it only executes during a time-to-close delay (so for a 10 second delay, it is called 20 times) to cycle the light and on completion of the timeout close the door.

[the remark this answers is not in the document]
Comment (dkerr64) [reply src/comms.cpp:896]: > Right now in the transmitSec2 function we write the rolling code to flash every time. I think we need to add something to only write once an hour or something. This should be ok now that the device is stable enough that it stays up for days at a time. The esphome variant does something along these lines.
> 
> I think we have to do this now because closing the garage door now will trigger a lot more writes to flash.

If we do this (which makes sense) then we need some sort of error recovery... as we now have EspSaveCrash with a callback then maybe we have to write to the flash whenever it crashes.  And is there a way to detect on boot-up that things are out-of-sync and re-initialize it... then I can remove the code file deletes from protocol change.

File: src/web.cpp (hunk 1 above)
Comment (dkerr64) [reply src/web.cpp]: True, but no harm to leave it.  Is useful I think.

File: src/web.cpp (hunk 2 above)
Comment (dkerr64) [reply src/web.cpp:106]: I suspect because it has C++ classes within the struct.  And those have memory malloc/free behind the scenes.

sizeof(SSESubscription) is 216 bytes (according to VScode, and with my logViewer bool), so array of 4 is 864 bytes.  I think limiting to 4 is okay.

File: src/comms.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (jgstroud) [reply src/comms.cpp:896]: for reference: https://github.com/ratgdo/esphome-ratgdo/blob/main/components/ratgdo/secplus2.cpp#L25

[the remark this answers is not in the document]
Comment (dkerr64) [reply src/comms.cpp:896]: Ah, so the GDO is tolerant of incrementing >1 in the rolling code, and ESPhome guys took a guess at 60.  We could do the same, but note that time-to-close light flash sends two per second, so our math might be a little different... if we write once a minute, then in theory we could have >120 code increments.  And we can also handle the crash case with our callback, and manually requested reboots.  Which just leaves power outages as uncaught.

[the remark this answers is not in the document]
Comment (dkerr64) [reply src/comms.cpp:896]: btw... what is the expected life of the ESP8266 flash?

**Comment by dkerr64** [comment]
I rebased onto latest main. Sorry for several force-push.  Had several conflicts in web.cpp to get right.

I've been running with a browser open for 17 hours now, with server logs going to javascript console. All good.

I was originally thinking of having a separate URL for a separate browser to connect to, and for messages to appear in that browser.  But when I looked into it, I decided that camping onto the existing server-sent-event code was easiest and most efficient on ESP memory usage, and just dumping to the browser's javascript console was much simpler than anything else.  The server side code is very efficient.  Just don't put any RINFO() inside the code that processes it, else infinite loop!!

I added user setting to turn this on, that setting is local to the browser (localStorage) and I hide it if you are on mobile device as makes no sense to send logs to those.

Last piece is to change HomeKit library to capture its logs.

**Comment by jgstroud** [comment]
The rebase and force push is fine.  I haven't had time to test this out yet. Sounds like the new memory layout option isn't showing any negative effects yet. How is your free heap looking?  Can multiple clients be monitoring the serial log?

**Comment by dkerr64** [comment]
No negative effects.  But we will have to keep an eye on things. Free heap is in good condition.

image

As for the IRAM heap.  We need to keep track...
- json string buffer... 1024 bytes.
- log lineBuffer... 1024 bytes (could probably reduce to 512 or less, for sprintf() into for each long line.
- log msgBuffer... 2048 bytes (holds about last 20 log lines, could increase to 4096?).
- savedLogs... 2048 (malloc's and free'd to read saved log file into when display requested).

In other words, we need at least 2x the message buffer space as we must allocate space to read the saved file into.  Now in theory I could probably read the file in chunks, but it would get rather complicated because the "head" of the circular buffer could be anywhere in the file.  So long as we have the space, I prefer reading it all in at once.

**Comment by jgstroud** [comment]
We can log the free IRAM heap as well, but maybe not add to web ui  

https://github.com/esp8266/Arduino/blob/master/libraries/esp8266/examples/irammem/irammem.ino#L465

**Comment by dkerr64** [comment]
> We can log the free IRAM heap as well, but maybe not add to web ui
> 
> https://github.com/esp8266/Arduino/blob/master/libraries/esp8266/examples/irammem/irammem.ino#L465

I don't think we need to.  It is very stable, we only malloc on initialization.  I just removed the dynamic malloc/free when reading from the saved file and instead malloc that buffer on initialization like the rest.  So free space will never change during runtime.

**Comment by dkerr64** [comment]
I added HomeKit library logs to our custom logging.

This is ready for a thorough thrashing and subsequent merging.  I don't have anything left on my to-do list for now.

**Comment by dkerr64** [comment]
@jgstroud just remembered... do you want to put back in the setTimeout(500).  I'll leave that to you to decide.

**Comment by dkerr64** [comment]
And also... you might want to implement a rolling code write-to-flash minimizing solution.  If you do, suggest you write it on crash callback and on user requested reboots.

**Comment by jgstroud** [comment]
Just leave it as is for now.  I did a quick review of the code.  Once I get a chance to test it, we can merge.  But I'll probably want to run this code for a week or so before we release with it.

**Comment by jgstroud** [comment]
> And also... you might want to implement a rolling code write-to-flash minimizing solution.  If you do, suggest you write it on crash callback and on user requested reboots.


Yes I definitely want to add this before we release.

**Review by jgstroud — COMMENTED** [review]
Do you still want the [Keep track of last door state change](https://github.com/ratgdo/homekit-ratgdo/pull/145/commits/97013c0abe18f36d31e30e52d37f1c46fe0fda94) commit without the NTP client

**Comment by jgstroud** [comment]
I've rebased, reordered some of the commits and squashed some commits.  I don't think I can update this PR, so I'm going to replace this PR with a new one

**Comment by dkerr64** [comment]
> Do you still want the [Keep track of last door state change](https://github.com/ratgdo/homekit-ratgdo/pull/145/commits/97013c0abe18f36d31e30e52d37f1c46fe0fda94) commit without the NTP client

Yes please, I'd like to keep that.  I removed all the NTP client so it just tracks it from last reboot.  I find it useful and its minor stuff.

**Comment by dkerr64** [comment]
> I've rebased, reordered some of the commits and squashed some commits. I don't think I can update this PR, so I'm going to replace this PR with a new one

Yes, definitely some squashing can be done.  Thanks.  I can make you a collaborator (thought you already were) or you can do a new PR if you prefer.  Your choice.

**Comment by jgstroud** [comment]
> > I've rebased, reordered some of the commits and squashed some commits. I don't think I can update this PR, so I'm going to replace this PR with a new one
> 
> Yes, definitely some squashing can be done. Thanks. I can make you a collaborator (thought you already were) or you can do a new PR if you prefer. Your choice.

I was able to update the PR. Thanks.   It could probably be cleaned up more, but for the most part, I'm happy with this

**Comment by jgstroud** [comment]
I never moved the crashlog from eeprom to filesystem.   It started to seem like a big rewrite of the utility and I'm not sure we gain a whole lot from it.  we can do that later though

**Comment by dkerr64** [comment]
> I never moved the crashlog from eeprom to filesystem. It started to seem like a big rewrite of the utility and I'm not sure we gain a whole lot from it. we can do that later though

I'm fine with that.  What about overwriting previous crash data, so it only attempts to keep the latest?  Save that for future too?

**Comment by dkerr64** [comment]
> > > I've rebased, reordered some of the commits and squashed some commits. I don't think I can update this PR, so I'm going to replace this PR with a new one
> > 
> > 
> > Yes, definitely some squashing can be done. Thanks. I can make you a collaborator (thought you already were) or you can do a new PR if you prefer. Your choice.
> 
> I was able to update the PR. Thanks. It could probably be cleaned up more, but for the most part, I'm happy with this

good, thanks.  Embarrisingly I still have that _missing semicolon_ commit visible to all !!!  But I suppose that demonstrates integrity and that I'm not ashamed to make my trivial errors visible. :-)

**Comment by jgstroud** [comment]
> > > > I've rebased, reordered some of the commits and squashed some commits. I don't think I can update this PR, so I'm going to replace this PR with a new one
> > > 
> > > 
> > > Yes, definitely some squashing can be done. Thanks. I can make you a collaborator (thought you already were) or you can do a new PR if you prefer. Your choice.
> > 
> > 
> > I was able to update the PR. Thanks. It could probably be cleaned up more, but for the most part, I'm happy with this
> 
> good, thanks. Embarrassingly I still have that _missing semicolon_ commit visible to all !!! But I suppose that demonstrates integrity and that I'm not ashamed to make my trivial errors visible. :-)

oops, missed that one
